parsel/selector.py

```diff
@@ -319,6 +399,12 @@ def xpath(
 
             selector.xpath('//a[href=$url]', url="http://www.example.com")
         """
+        if self.type == "text":
+            self._load_lxml_root(self.root, type="html")
```

Comment:
Is it a kind of lazy loading? I wonder if we can implement it in some other way. It's not great that .xpath method mutates the state of the Selector. It seems selector.root and selector.type can be different before and after calling .xpath method.

Reply:
E.g. `sel.jmespath(...)` is going to behave differently if `sel.xpath(...)` was called before.

Reply:
Yes, @felipeboffnunes mention this to me privately as well, though I dismissed it.

I do not think it is a big deal given that valid HTML/XML cannot be valid JSON. As long as we modify `_load_lxml_root` so that the change to `self.type` happens _after_ parsing the text, and assuming the parsing of invalid HTML/XML causes an exception, the problem you mention should be gone, I think.

However, if we can avoid the `type` reassignment without changing the API (e.g. still allowing cross-input-type chaining without explicitly passing a `type` parameter), I guess it is a better approach.

Reply:
Hey! This thing is my largest concern. It'd be much better if there is no mutation of state in selectors, and I think we should fix this before merging this PR. My other comments are more like suggestions, I'm fine with proceeding without addressing all of them.

Reply:
https://github.com/scrapy/parsel/pull/181/commits/812b2bf05b9c29827efb4a784cf37525f0e1132f